docs/eta-pi-p/manual.ipynb

```diff
@@ -255,24 +276,235 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Visualization"
+    "### Kinematic variables"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@unevaluated\n",
+    "class SquaredInvariantMass(sp.Expr):\n",
+    "    momentum: sp.Basic\n",
+    "    _latex_repr_ = \"m_{{{momentum}}}^2\"\n",
+    "\n",
+    "    def evaluate(self) -> sp.Expr:\n",
+    "        p = self.momentum\n",
+    "        p_xyz = ThreeMomentum(p)\n",
+    "        return Energy(p) ** 2 - EuclideanNorm(p_xyz) ** 2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "jupyter": {
+     "source_hidden": true
+    }
+   },
+   "outputs": [],
+   "source": [
+    "p1 = FourMomentumSymbol(\"p1\", shape=[])\n",
+    "p2 = FourMomentumSymbol(\"p2\", shape=[])\n",
+    "p3 = FourMomentumSymbol(\"p3\", shape=[])\n",
+    "p12 = ArraySum(p1, p2)\n",
+    "p23 = ArraySum(p2, p3)\n",
+    "p31 = ArraySum(p3, p1)\n",
+    "\n",
+    "theta12 = Theta(p12)\n",
+    "theta23 = Theta(p23)\n",
+    "theta31 = Theta(p31)\n",
+    "\n",
+    "phi12 = Phi(p12)\n",
+    "phi23 = Phi(p23)\n",
+    "phi31 = Phi(p31)\n",
+    "\n",
+    "beta12 = three_momentum_norm(p12) / Energy(p12)\n",
+    "beta23 = three_momentum_norm(p23) / Energy(p23)\n",
+    "beta31 = three_momentum_norm(p31) / Energy(p31)\n",
+    "\n",
+    "Bz12 = BoostZMatrix(beta12, n_events=ArraySize(beta12))\n",
+    "Bz23 = BoostZMatrix(beta23, n_events=ArraySize(beta23))\n",
+    "Bz31 = BoostZMatrix(beta31, n_events=ArraySize(beta31))\n",
+    "\n",
+    "Ry12 = RotationYMatrix(-theta12, n_events=ArraySize(theta12))\n",
+    "Ry23 = RotationYMatrix(-theta23, n_events=ArraySize(theta23))\n",
+    "Ry31 = RotationYMatrix(-theta31, n_events=ArraySize(theta31))\n",
+    "\n",
+    "Rz12 = RotationZMatrix(-phi12, n_events=ArraySize(theta12))\n",
+    "Rz23 = RotationZMatrix(-phi23, n_events=ArraySize(theta23))\n",
+    "Rz31 = RotationZMatrix(-phi31, n_events=ArraySize(theta31))\n",
+    "\n",
+    "m12_square = SquaredInvariantMass(p12)\n",
+    "m23_square = SquaredInvariantMass(p23)\n",
+    "m31_square = SquaredInvariantMass(p31)\n",
+    "\n",
+    "phi12_helicity = Phi(ArrayMultiplication(Bz12, Ry12, Rz12, p1))\n",
+    "phi23_helicity = Phi(ArrayMultiplication(Bz23, Ry23, Rz23, p2))\n",
+    "phi31_helicity = Phi(ArrayMultiplication(Bz31, Ry31, Rz31, p3))\n",
+    "\n",
+    "theta12_helicity = Theta(ArrayMultiplication(Bz12, Ry12, Rz12, p1))\n",
+    "theta23_helicity = Theta(ArrayMultiplication(Bz23, Ry23, Rz23, p2))\n",
+    "theta31_helicity = Theta(ArrayMultiplication(Bz31, Ry31, Rz31, p3))"
```

Comment:
There's a lot of repetitive code here. I would create one function that formulate the angle directly and does all the
boosting steps inside, e.g.:
```python
def formulate_helicity_angles(pi: FourMomentumSymbol, pj: FourMomentumSymbol) -> tuple[Theta, Phi]:
    ...
    return Phi(pi_boosted), Theta(pi_boosted)
```

Or you define one rotated boost function:

```python
def formulate_rotated_boost(p: sp.Basic, boost_p: sp.Basic) -> ArrayMultiplication:
    return ArrayMultiplication(Bz, Ry, Rz, p)


formulate_rotated_boost(p1, p12)
```

Reply:
nice tip! 🍩